src/helpers/callback-proxy.ts

```diff
@@ -0,0 +1,65 @@
+import { issueCommentCreatedCallback } from "../handlers/comment-created-callback";
+import { Context, SupportedEventsU } from "../types";
+import { ProxyCallbacks } from "../types/proxy";
+import { bubbleUpErrorComment } from "./errors";
+
+/**
+ * The `callbacks` object defines an array of callback functions for each supported event type.
+ *
+ * Since multiple callbacks might need to be executed for a single event, we store each
+ * callback in an array. This design allows for extensibility and flexibility, enabling
+ * us to add more callbacks for a particular event without modifying the core logic.
+ */
+const callbacks = {
+  "issue_comment.created": [issueCommentCreatedCallback],
+} as ProxyCallbacks;
+
+/**
+ * The `proxyCallbacks` function returns a Proxy object that intercepts access to the
+ * `callbacks` object. This Proxy enables dynamic handling of event callbacks, including:
+ *
+ * - **Event Handling:** When an event occurs, the Proxy looks up the corresponding
+ *   callbacks in the `callbacks` object. If no callbacks are found for the event,
+ *   it returns a `skipped` status.
+ *
+ * - **Error Handling:** If an error occurs while processing a callback, the Proxy
+ *   logs the error and returns a `failed` status.
+ *
+ * The Proxy uses the `get` trap to intercept attempts to access properties on the
+ * `callbacks` object. This trap allows us to asynchronously execute the appropriate
+ * callbacks based on the event type, ensuring that the correct context is passed to
+ * each callback.
+ */
+export function proxyCallbacks(context: Context): ProxyCallbacks {
+  return new Proxy(callbacks, {
+    get(target, prop: SupportedEventsU) {
+      if (!target[prop]) {
+        context.logger.info(`No callbacks found for event ${prop}`);
+        return { status: 204, reason: "skipped" };
+      }
+      return (async () => {
+        try {
+          return await Promise.all(target[prop].map((callback) => handleCallback(callback, context)));
```

Comment:
wouldn't `Promise.allSettled` make more sense because `.all` will fail even if the first one fails...well I guess it's more how you design your callbacks, sometimes 2 callbacks are dependant on each other so you want first to succeed to run the second one, but sometimes you want them independent

Reply:
Hmm idk you let me know what to do in this scenario mate. Even though it stops others potentially I think it might be better to stick with `.all`.


---

On the matter of callbacks, we had some misunderstanding with how to properly catch the errors with the proxy you might recall. Well this implementation seems to be the way to handle it.

`runPlugin` returns `proxyCallbacks(context)[eventName]` no try-catch. The Proxy object will post the comment inside it's catch using `bubbleupErrorComment`. Which means callbacks need to either just throw their error or use `bubbleupErrorComment` with the posting param disabled. I've found this to work well so far.

Reply:
We can leave it for now, it's not that important.
I wonder how this Proxy approach will work with the SDK because with the SDK you provide a function that the SDK will call and the SDK will catch the error and post a comment

Reply:
I liked your Hono worker approach when I first saw it, is this side of the SDK ready to be used to build a plugin yet, I can try it with https://github.com/ubiquity-os/permit-generation/issues/92#event-14901391331?

`proxyCallbacks(ctx)` returns the callback obj
`proxyCallbacks(ctx)[eventName]` returns a single async function which loops all callbacks for that event in the obj. 

So I think we can treat each event as a single `handler` and you'd be in control of `.all` or `.allSettled` if you have the user pass in the callbacks obj to the `proxyCallbacks` function, currently we pull it from file scope

The fn/promise returned by `proxyCallbacks(ctx)[eventName]` posts the error inside it's catch block rn but if it was thrown in there I'm sure the invoking fn would need to catch it which would be the SDK. Needs confirmed tho as that may have been our problem in the first place, throwing it inside the promise and the proxy interfered it actually throwing it. 

I think we should aim to standardize and support `proxyCallbacks` out of the box because it's the best method found so far to handle every webhook payload at the type-level elegantly.


```
npx tsx t.ts
proxyCallbacks {
  'issue_comment.created': [ [AsyncFunction: notificationsRequiringComments] ],
  'issue_comment.edited': [ [AsyncFunction: notificationsRequiringComments] ],
  'issues.unassigned': [ [AsyncFunction: disqualificationNotification] ],
  'pull_request.review_requested': [ [AsyncFunction: reviewNotification] ]
}
```
```
npx tsx t.ts
proxyCallbacks Promise { <pending> }
```